src/com/facebook/buck/swift/SwiftCompile.java

```diff
@@ -265,15 +324,97 @@ public Path getPathToOutput() {
     return args.build();
   }
 
-  ImmutableSet<Arg> getLinkArgs() {
+  ImmutableSet<Arg> getAstLinkArgs() {
     return ImmutableSet.<Arg>builder()
         .addAll(StringArg.from("-Xlinker", "-add_ast_path"))
         .add(new SourcePathArg(
             getResolver(),
             new BuildTargetSourcePath(getBuildTarget(), modulePath)))
-        .add(new SourcePathArg(
-            getResolver(),
-            new BuildTargetSourcePath(getBuildTarget(), objectPath)))
         .build();
   }
+
+  ImmutableList<Arg> getFileListLinkArgs() {
+    return FileListableLinkerInputArg.from(
+        FluentIterable.from(SourcePathArg.from(getResolver(), getObjectLists()))
+            .filter(SourcePathArg.class)
+            .toList());
+  }
+
+  private ImmutableList<SourcePath> getObjectLists() {
+    return FluentIterable.from(getDeclaredDeps())
+        .filter(SwiftCompile.class)
+        .filter(IS_COMPILE_INDIVIDUAL)
+        .transform(new Function<SwiftCompile, SourcePath>() {
+          @Override
+          public SourcePath apply(SwiftCompile input) {
+            return new BuildTargetSourcePath(getBuildTarget(), input.objectPath);
+          }
+        })
+        .toList();
+  }
+
+  public static BuildRule of(
+      CxxPlatform cxxPlatform,
+      SwiftBuckConfig swiftBuckConfig,
+      BuildRuleParams params,
+      BuildRuleResolver resolver,
+      SourcePathResolver sourcePathResolver,
+      Tool swiftCompiler,
+      ImmutableSortedSet<FrameworkPath> frameworkPaths,
+      String moduleName,
+      Path outputPath,
+      ImmutableSet<SourcePath> srcs,
+      Optional<Boolean> enableObjcInterop,
+      Optional<SourcePath> bridgingHeader) throws NoSuchBuildTargetException {
+    SwiftPrepareForCompile prepareForCompile = new SwiftPrepareForCompile(
+        params.withFlavor(PREPARE_FLAVOR),
+        sourcePathResolver,
+        outputPath,
+        srcs);
+    params = params.appendExtraDeps(resolver.addToIndex(prepareForCompile));
+
+    ImmutableSortedSet.Builder<BuildRule> deps = ImmutableSortedSet.naturalOrder();
+    deps.addAll(params.getDeclaredDeps().get());
+
+    // create swift compile for every single file
+    for (SourcePath src: srcs) {
+      String outputName = CxxFlavorSanitizer.sanitize(
+          sourcePathResolver.getAbsolutePath(src).getFileName() + ".o");
+      BuildRule compileSingleFile =
+          new SwiftCompile(
+              cxxPlatform,
+              swiftBuckConfig,
+              params.withFlavor(
+                  ImmutableFlavor.of(
+                      COMPILE_FLAVOR_PREFIX + outputName)),
+              sourcePathResolver,
+              swiftCompiler,
+              frameworkPaths,
+              moduleName,
+              outputPath,
```

Comment:
should we put all individual compiled file to the same folder like this, or use separate folders like cxx precompile?
I thought it could be a race condition when you compile `file1.swift` and `file2.swift` in parallel given `file2.swift` depending on `file1.swift`, but apparently compiling `file2.swift` doesn't result to compiling `file1.swift` (maybe it will do parsing & semantic analysis but won't actually compile `file1.swift`?). This is a really interesting topic and we need to understand more about swift compiler.